### PR TITLE
Use constexpr statements over enable_if dispatched functions

### PR DESCRIPTION
**Pull request by blowekamp**
The conventional if statements with "constexpr" is more readable than complicated enable_if statements. Additionally, the produced object files are smaller due to shorter symbols names.

### CODE REVIEW

**Review by zivy — APPROVED** [review]

File: Code/IO/include/sitkImportImageFilter.h [hunk 1]
```diff
@@ -118,11 +118,7 @@ class SITKIO_EXPORT ImportImageFilter : public ImageReaderBase
   // components per pixel needs to be set, otherwise the method
   // does not exist. This is done with the enable if idiom.
   template <class TImageType>
-  typename std::enable_if<!IsVector<TImageType>::Value>::type
-  SetNumberOfComponentsOnImage(TImageType *)
-  {}
-  template <class TImageType>
-  typename std::enable_if<IsVector<TImageType>::Value>::type
+  void
   SetNumberOfComponentsOnImage(TImageType *);
```
Comment (N-Dekker) [inline Code/IO/include/sitkImportImageFilter.h:122]: Looks like this member function declaration (`SetNumberOfComponentsOnImage`) may be removed as well!

File: Code/Common/include/sitkMemberFunctionFactoryBase.h [hunk 2]
```diff
@@ -69,21 +69,16 @@ template <class... TupleArgs>
...
-    inline typename std::enable_if < Idx<sizeof...(TupleTypes), void>::type
-                                     hash_combine_tup(size_t & seed, const std::tuple<TupleTypes...> & tup) const
+  void
+  hash_combine_tup(size_t & seed, const std::tuple<TupleTypes...> & tup) const
   {
     hash_combine(seed, std::get<Idx>(tup));
 
-    //  on to next element
-    hash_combine_tup<Idx + 1>(seed, tup);
+    if constexpr (Idx + 1< sizeof...(TupleTypes))
```
Comment (blowekamp) [inline Code/Common/include/sitkMemberFunctionFactoryBase.h:78]: @N-Dekker This was one of the most interesting replacements on a recursive variadic template function.

Comment (N-Dekker) [reply Code/Common/include/sitkMemberFunctionFactoryBase.h:78]: @blowekamp Nice, but maybe it can even be further improved! Let me think...

Comment (N-Dekker) [reply Code/Common/include/sitkMemberFunctionFactoryBase.h:78]: (hash_combine(seed, std::get<Idx>(tup), ...)

Comment (N-Dekker) [reply Code/Common/include/sitkMemberFunctionFactoryBase.h:78]: Using a C++17 [fold expression](https://en.cppreference.com/w/cpp/language/fold) and [std::apply](https://en.cppreference.com/w/cpp/utility/apply)...?

```cpp
  operator()(std::tuple<TupleArgs...> tupleValue) const
  {
    size_t seed = 0;
    std::apply(
      [this, &seed](auto... tupleElement) { (hash_combine(seed, tupleElement), ...); },
      tupleValue);
    return seed;
  }
```

Disclaimer: uncompiled, untested code 🙀

Comment (blowekamp) [reply Code/Common/include/sitkMemberFunctionFactoryBase.h:78]: Yes, I just stumble onto std::apply now too. However, I think just: `std::apply(hash_combine,tupleValue);` is what is needed 🤔

**Comment by N-Dekker** [comment]
@blowekamp I'm sorry I don't have time to review the entire PR thoroughly, but you don't need to wait for me. I very much like its direction, and I'm glad I could help with a few suggestions 👍

File: Code/Common/include/sitkMemberFunctionFactoryBase.h (hunk 2 above)
Comment (blowekamp) [reply Code/Common/include/sitkMemberFunctionFactoryBase.h]: Thank you very much for this suggestion.

**Pull request merged**